raft/src/main/java/org/apache/kafka/raft/internals/PartitionListener.java

```diff
@@ -0,0 +1,259 @@
+     */
+    public short kraftVersionAtOffset(long offset) {
+        long fixedNextOffset = nextOffset;
+        if (offset >= fixedNextOffset) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Attempting the read the kraft.version at an offset (%d) which kraft hasn't seen (%d)",
+                    offset,
+                    fixedNextOffset - 1
+                )
+            );
+        }
+
+        synchronized (kraftVersionHistory) {
+            return kraftVersionHistory.valueAtOrBefore(offset).orElse((short) 0);
+        }
+    }
+
+    private void maybeLoadLog() {
+        while (log.endOffset().offset > nextOffset) {
+            LogFetchInfo info = log.read(nextOffset, Isolation.UNCOMMITTED);
+            try (RecordsIterator<?> iterator = new RecordsIterator<>(
+                    info.records,
+                    serde,
+                    bufferSupplier,
+                    maxBatchSizeBytes,
+                    true // Validate batch CRC
+                )
+            ) {
+                while (iterator.hasNext()) {
+                    Batch<?> batch = iterator.next();
+                    handleBatch(batch, OptionalLong.empty());
+                    nextOffset = batch.lastOffset() + 1;
+                }
+            }
+        }
+    }
+
+    private void maybeLoadSnapshot() {
+        if ((nextOffset == 0 || nextOffset < log.startOffset()) && log.latestSnapshot().isPresent()) {
+            RawSnapshotReader rawSnapshot = log.latestSnapshot().get();
+            // Clear the current state
+            synchronized (kraftVersionHistory) {
+                kraftVersionHistory.clear();
+            }
+            synchronized (voterSetHistory) {
+                voterSetHistory.clear();
+            }
+
+            // Load the snapshot since the listener is at the start of the log or the log doesn't have the next entry.
+            try (SnapshotReader<?> reader = RecordsSnapshotReader.of(
+                    rawSnapshot,
+                    serde,
+                    bufferSupplier,
+                    maxBatchSizeBytes,
+                    true // Validate batch CRC
+                )
+            ) {
+                logger.info(
+                    "Loading snapshot ({}) since log start offset ({}) is greater than the internal listener's next offset ({})",
+                    reader.snapshotId(),
+                    log.startOffset(),
+                    nextOffset
+                );
+                OptionalLong currentOffset = OptionalLong.of(reader.lastContainedLogOffset());
+                while (reader.hasNext()) {
+                    Batch<?> batch = reader.next();
+                    handleBatch(batch, currentOffset);
+                }
+
+                nextOffset = reader.lastContainedLogOffset() + 1;
+            }
+        }
+    }
+
+    private void handleBatch(Batch<?> batch, OptionalLong overrideOffset) {
+        int index = 0;
```

Comment:
I wonder if it is better to get the record offsets from the `Batch` object. This works because we know the partition is not compacted, but it still feels a bit suspicious.

Reply:
The issue is that the offset in the batch are only meaningful when handling batches that come from the log. The offset of batches that come from the snapshot don't have any relation to the log offset. This is why we pass `OptionalLong overrideOffset`.

I change this name to `offsetDelta`.